go/utils/intoto/slsa_zero_two_converter.go

```diff
@@ -0,0 +1,131 @@
+package intoto
+
+import (
+	"fmt"
+	"time"
+
+	pb "github.com/grafeas/grafeas/proto/v1/grafeas_go_proto"
+	intoto "github.com/in-toto/in-toto-golang/in_toto"
+	"google.golang.org/protobuf/types/known/structpb"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// SLSAZeroTwoToProto function converts the intoto statement with SLSA 0.2 predicate to the
+// grafeas proto InTotoStatement where slsa 0.2 predicate is oneof the predicate types
+func SLSAZeroTwoToProto(in *intoto.ProvenanceStatement) (*pb.InTotoStatement, error) {
+	if in == nil {
+		return nil, nil
+	}
+
+	subject := convertSubject(in.Subject)
+	predicate, err := convertPredicate(in)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.InTotoStatement{
+		Type:          in.Type,
+		Subject:       subject,
+		PredicateType: in.PredicateType,
+		Predicate:     predicate,
+	}, nil
+
+}
+
+func convertPredicate(in *intoto.ProvenanceStatement) (*pb.InTotoStatement_SlsaProvenanceZeroTwo, error) {
+	// convert the fields of interface{} type to *structpb.Struct
+	parameter, err := convertEmptyInterfaceToProtoStruct(in.Predicate.Invocation.Parameters)
+	if err != nil {
+		return nil, err
+	}
+
+	buildConfig, err := convertEmptyInterfaceToProtoStruct(in.Predicate.BuildConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	environment, err := convertEmptyInterfaceToProtoStruct(in.Predicate.Invocation.Environment)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.InTotoStatement_SlsaProvenanceZeroTwo{
+		SlsaProvenanceZeroTwo: &pb.SlsaProvenanceZeroTwo{
+			Builder:   &pb.SlsaProvenanceZeroTwo_SlsaBuilder{Id: in.Predicate.Builder.ID},
+			BuildType: in.Predicate.BuildType,
+			Invocation: &pb.SlsaProvenanceZeroTwo_SlsaInvocation{
+				ConfigSource: &pb.SlsaProvenanceZeroTwo_SlsaConfigSource{
+					Uri:        in.Predicate.Invocation.ConfigSource.URI,
+					Digest:     in.Predicate.Invocation.ConfigSource.Digest,
+					EntryPoint: in.Predicate.Invocation.ConfigSource.EntryPoint,
+				},
+				Parameters:  parameter,
+				Environment: environment,
+			},
+			BuildConfig: buildConfig,
+			Metadata:    convertMetaData(in),
+			Materials:   convertMaterials(in),
+		},
+	}, nil
+}
+
+func convertSubject(input []intoto.Subject) []*pb.Subject {
+	results := []*pb.Subject{}
+	for _, subject := range input {
+		results = append(results, &pb.Subject{
+			Name:   subject.Name,
+			Digest: subject.Digest,
+		})
+	}
+	return results
+}
+
+func convertEmptyInterfaceToProtoStruct(input interface{}) (*structpb.Struct, error) {
+	if input == nil {
+		return nil, nil
+	}
+
+	data, err := structpb.NewStruct(input.(map[string]interface{}))
```

Comment:
I think this panics if the incoming type is not `map[string]interface{}` - you may need to serialize the input -> map[string]interface{} first to make sure this works for all types.

Reply:
(we should also add a test for this)

Reply:
Since [structpb.struct](https://pkg.go.dev/google.golang.org/protobuf/types/known/structpb#Struct) only expects a map rather than any other type i.e string, slice, the actual type of the incoming interface{} has to be a map right?

Just changed `toProtoStruct` function in which we do marshal -> unmarshal to structpb.struct so that error will be returned instead of panic when the incoming type is not map[string]interface{}. Let me know what you think.

Reply:
btw, I am wondering if there is a proto type that can be autotranslated to go's `interface{}` type?

cc @sherzberg-1 since [build_config](https://github.com/grafeas/grafeas/blob/bc2e0363cdefcb36f4f5fa20bf45e7ff9f588030/proto/v1/slsa_provenance_zero_two.proto#L79), [parameters and environment](https://github.com/grafeas/grafeas/blob/bc2e0363cdefcb36f4f5fa20bf45e7ff9f588030/proto/v1/slsa_provenance_zero_two.proto#L47-L48) fields are designed to be structpb.Struct type.